File: 2342. Max Sum of a Pair With Equal Sum of Digits/main.py
from collections import defaultdict
import heapq
from typing import List


class Solution:
    def maximumSum(self, nums: List[int]) -> int:
        groups = defaultdict(list)

        def get_integer_sum(n):
            res = 0
            for c in str(n):
                res += int(c)
            return res
        
        for n in nums:
            heapq.heappush(groups[get_integer_sum(n)], -n)
        
        res = 0

        for values in groups.values():
            if len(values) >= 2:
                x, y = -heapq.heappop(values), -heapq.heappop(values)
                res = max(res, x + y)

        return res if res > 0 else -1
        

File: 2342. Max Sum of a Pair With Equal Sum of Digits/main.go
package main

import (
	"container/heap"
)

type MaxHeap []int

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x any) {
	*h = append(*h, x.(int))
}

func (h *MaxHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func maximumSum(nums []int) int {
	groups := make(map[int]*MaxHeap)

	getIntegerSum := func(n int) int {
		res := 0
		for n > 0 {
			res += n % 10
			n /= 10
		}
		return res
	}

	for _, n := range nums {
		group := getIntegerSum(n)
		if groups[group] == nil {
			groups[group] = &MaxHeap{}
		}
		heap.Push(groups[group], n)
	}

	res := 0

	for _, items := range groups {
		if items.Len() < 2 {
			continue
		}
		x, y := heap.Pop(items), heap.Pop(items)
		res = max(res, x.(int)+y.(int))
	}

	if res == 0 {
		return -1
	}

	return res
}
